File: core/src/main/java/vn/loitp/views/draggableflipview/FlipListener.java
package vn.loitp.views.draggableflipview;

import android.animation.ValueAnimator;
import android.view.View;

/**
 * this is animation class for rotate
 *
 * Created by sasakicks on 2015/09/09.
 */
public class FlipListener implements ValueAnimator.AnimatorUpdateListener {

    private View mParentView;
    private View mFrontView;
    private View mBackView;
    private boolean mFlipped;
    private int mDirection;

    public FlipListener(final View front, final View back, final View parent) {
        mParentView = parent;
        mFrontView = front;
        mBackView = back;
        mBackView.setVisibility(View.GONE);
    }

    @Override
    public void onAnimationUpdate(final ValueAnimator animation) {
        final float value = animation.getAnimatedFraction();
        final float scaleValue = 0.625f + (1.5f * (value - 0.5f) * (value - 0.5f));

        if (value <= 0.5f) {
            mParentView.setRotationY(180 * value * mDirection);
            if (mFlipped) setStateFlipped(false);
        } else {
            mParentView.setRotationY(-180 * (1 - value) * mDirection);
            if (!mFlipped) setStateFlipped(true);
        }
        mParentView.setScaleX(scaleValue);
        mParentView.setScaleY(scaleValue);
    }

    public void reverse() {
        View temp = mBackView;
        mBackView = mFrontView;
        mFrontView = temp;
    }

    public void setRotateDirection(int direction) {
        mDirection = direction;
    }

    private void setStateFlipped(boolean flipped) {
        mFlipped = flipped;
        mFrontView.setVisibility(flipped ? View.GONE : View.VISIBLE);
        mBackView.setVisibility(flipped ? View.VISIBLE : View.GONE);
    }
}

File: core/src/main/java/com/loitp/views/draggableFlipView/FlipListener.kt
package com.loitp.views.draggableFlipView

import android.animation.ValueAnimator
import android.view.View

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class FlipListener(
    private var mFrontView: View?,
    private var mBackView: View?,
    private val mParentView: View
) : ValueAnimator.AnimatorUpdateListener {

    private var mFlipped = false
    private var mDirection = 0

    override fun onAnimationUpdate(animation: ValueAnimator) {
        val value = animation.animatedFraction
        val scaleValue = 0.625f + 1.5f * (value - 0.5f) * (value - 0.5f)
        if (value <= 0.5f) {
            mParentView.rotationY = 180 * value * mDirection
            if (mFlipped) setStateFlipped(false)
        } else {
            mParentView.rotationY = -180 * (1 - value) * mDirection
            if (!mFlipped) setStateFlipped(true)
        }
        mParentView.scaleX = scaleValue
        mParentView.scaleY = scaleValue
    }

    fun reverse() {
        val temp = mBackView
        mBackView = mFrontView
        mFrontView = temp
    }

    fun setRotateDirection(direction: Int) {
        mDirection = direction
    }

    private fun setStateFlipped(flipped: Boolean) {
        mFlipped = flipped
        mFrontView?.visibility = if (flipped) View.GONE else View.VISIBLE
        mBackView?.visibility = if (flipped) View.VISIBLE else View.GONE
    }

    init {
        mBackView?.visibility = View.GONE
    }
}
